pkg/controller/tidbcluster/pod_control.go

```diff
@@ -0,0 +1,318 @@
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("TidbCluster pod: %v, sync failed %v, requeuing", key.(string), err))
+		c.queue.AddRateLimited(key)
+	} else {
+		if result.RequeueAfter > 0 {
+			c.queue.AddAfter(key, result.RequeueAfter)
+		} else if result.Requeue {
+			c.queue.AddRateLimited(key)
+		} else {
+			c.queue.Forget(key)
+		}
+	}
+	return true
+}
+
+func (c *PodController) sync(key string) (reconcile.Result, error) {
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	pod, err := c.deps.PodLister.Pods(ns).Get(name)
+	if errors.IsNotFound(err) {
+		klog.Infof("pod has been deleted %v", key)
+		return reconcile.Result{}, nil
+	}
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	pod = pod.DeepCopy()
+
+	// labels of tikv:
+	// app.kubernetes.io/component=tikv,app.kubernetes.io/instance=db1369682775200135879,app.kubernetes.io/managed-by=tidb-operator ...
+	managedBy := pod.Labels[label.ManagedByLabelKey]
+	if managedBy != label.TiDBOperator {
+		return reconcile.Result{}, nil
+	}
+
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing TidbCluster pod %q (%v)", key, time.Since(startTime))
+	}()
+
+	component := pod.Labels[label.ComponentLabelKey]
+	ctx := context.Background()
+	switch component {
+	case label.TiKVLabelVal:
+		tcName := pod.Labels[label.InstanceLabelKey]
+		tc, err := c.deps.TiDBClusterLister.TidbClusters(ns).Get(tcName)
+		if err != nil {
+			return reconcile.Result{}, perrors.Annotatef(err, "failed to get tc %q", ns+"/"+tcName)
+		}
+		tc = tc.DeepCopy()
+
+		return c.syncTiKVPod(ctx, pod, tc)
+	default:
+		return reconcile.Result{}, nil
+	}
+}
+
+func (c *PodController) getPDClient(tc *v1alpha1.TidbCluster) pdapi.PDClient {
+	if c.testPDClient != nil {
+		return c.testPDClient
+	}
+
+	pdClient := controller.GetPDClient(c.deps.PDControl, tc)
+	return pdClient
+}
+
+func (c *PodController) syncTiKVPod(ctx context.Context, pod *corev1.Pod, tc *v1alpha1.TidbCluster) (reconcile.Result, error) {
+	value, ok := pod.Annotations[v1alpha1.EvictLeaderAnnKey]
+
+	if ok {
+		switch value {
+		case v1alpha1.EvictLeaderValueNone:
+		case v1alpha1.EvictLeaderValueDeletePod:
+		default:
+			klog.Warningf("ignore unknown value %q of annotation %q", value, v1alpha1.EvictLeaderAnnKey)
+			return reconcile.Result{}, nil
+		}
+	}
+
+	if ok {
+		evictStatus := &v1alpha1.EvictLeaderStatus{
+			PodCreateTime: pod.CreationTimestamp,
+			Value:         value,
+		}
+		nowStatus := tc.Status.TiKV.EvictLeader[pod.Name]
+		if nowStatus == nil || *nowStatus != *evictStatus {
+			if tc.Status.TiKV.EvictLeader == nil {
+				tc.Status.TiKV.EvictLeader = make(map[string]*v1alpha1.EvictLeaderStatus)
+			}
+			tc.Status.TiKV.EvictLeader[pod.Name] = evictStatus
+			var err error
+			key := fmt.Sprintf("%s/%s", tc.Namespace, tc.Name)
+			tc, err = c.deps.Clientset.PingcapV1alpha1().TidbClusters(tc.Namespace).Update(ctx, tc, metav1.UpdateOptions{})
```

Comment:
Now we have both the Pod and tidb cluster controller update the same CR, it may cause conflict especially when the cluster controller updates the CR frequently, what about referring to the logic here https://github.com/pingcap/tidb-operator/blob/release-1.2/pkg/controller/tidbcluster_control.go#L63 to retry?

Example log:
Failed to update backup [default/minio-2021-12-07t05-14-00], error: Operation cannot be fulfilled on backups.pingcap.com "minio-2021-12-07t05-14-00": the object has been modified; please apply your changes to the latest version and try again

Reply:
https://github.com/pingcap/tidb-operator/blob/release-1.2/pkg/controller/tidbcluster_control.go#L63
after taking a look at `UpdateTidbCluster`
I think it has totally done things in the wrong way, it ignore conflict, it's like a force update the spec part.

example
1. tc v1: tikv.replicas = 2
2. operator get v1 and do some think...
3. I manual update tikv.replicas = 3.
4. operator force update it as the spec it wants to be(base on tc v1, so tikv.replicas = 2)


It lost my update as tikv.replicas = 3.

Reply:
`func (c *realTidbClusterControl) UpdateTidbCluster(tc *v1alpha1.TidbCluster, newStatus *v1alpha1.TidbClusterStatus, oldStatus *v1alpha1.TidbClusterStatus) (*v1alpha1.TidbCluster, error) {
`
https://github.com/pingcap/tidb-operator/blob/release-1.2/pkg/controller/tidbcluster_control.go#L63
WTF the params `newStatus` and `oldStatus` is useless

Reply:
here, I prefer to keep it simple by just returning conflict now, and it is always safe to reconcile.
generally, we will only add the annotation to restart when the cluster is not doing any operations(so tc not being changed).

Reply:
I put the code link just for your reference, not to use that directly.
BTW, `UpdateTidbCluster` works as expected because it only updates the `status` and the `spec` should not be updated by any controller.
> here, I prefer to keep it simple by just returning conflict now, and it is always safe to reconcile.
OK.
> generally, we will only add the annotation to restart when the cluster is not doing any operations(so tc not being changed).
Please remember that the `status` may be updated by the controller if something changes, and here we also update the `status`. BTW, users usually restart the Pod when there is something wrong with the Pod or the store.